src/librustc_middle/ty/layout.rs

```diff
@@ -285,9 +285,17 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
 
         let mut inverse_memory_index: Vec<u32> = (0..fields.len() as u32).collect();
 
+        let mut largest_niche_index = None;
         let mut optimize = !repr.inhibit_struct_field_reordering_opt();
         if let StructKind::Prefixed(_, align) = kind {
             optimize &= align.bytes() == 1;
```

Comment:
Why do we disable optimization if the discriminant need alignment?

Reply:
For discriminants larger than a byte (i.e. with alignment larger than 1), sorting by alignment won't produce the optimal layout, because we can't move the discriminant, and there might be fields with lower alignment in there. Therefore, we decide to let the user control the order instead.

Reply:
> Why do we need to "move the discriminant" ?

The intuition was that *all fields* (including the discriminant) had to be sorted by the alignment for optimal (i.e. minimal) size.

The interesting thing we hadn't realized, and I just spent some time thinking through, is that this sort of worst-case situation is still optimal:
* discriminant: `u32` (`align=4`)
* fields with `align<4`, total padding `0`, `1`, `2` or `3` bytes
* fields with `align>=4`

The first group of fields can be (conceptually) swapped with the discriminant, resulting in the sorted order that we know is optimal, but there's no way the padding could change, because they can't tell apart being at the start of the type or being after the discriminant (both are aligned to a multiple of `4`).

And the second group of fields starts on an offset that's a multiple of `4`, so they're not affected by the padding inside the first group of fields.

This has implications beyond `enum`s, e.g. `(u32, u8, u32, u64)` has minimal padding (`7`) and sorting is unnecessary - in fact, I can't think of any permutation of that tuple that would be larger than `24` bytes, [and we can test that](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5eb19ef9861e388f56bce93a762871c2).
However, I'm not sure if there's any point in doing anything clever based on that finding, outside of `enum`s.

Can you make a separate PR that removes `optimize &= align.bytes() == 1;` and adds some tests?

Reply:
> Can you make a separate PR that removes optimize &= align.bytes() == 1; and adds some tests?

https://github.com/rust-lang/rust/pull/70748